pinot-common/src/main/java/org/apache/pinot/common/utils/http/HttpClient.java

```diff
@@ -0,0 +1,371 @@
+      @Nullable String authToken)
+      throws IOException {
+    RequestBuilder requestBuilder = RequestBuilder.put(uri).setVersion(HttpVersion.HTTP_1_1);
+    if (payload != null) {
+      requestBuilder.setEntity(payload);
+    }
+    if (StringUtils.isNotBlank(authToken)) {
+      requestBuilder.addHeader("Authorization", authToken);
+    }
+    setTimeout(requestBuilder, DELETE_REQUEST_SOCKET_TIMEOUT_MS);
+    return sendRequest(requestBuilder.build());
+  }
+
+  // --------------------------------------------------------------------------
+  // JSON post/put utility APIs
+  // --------------------------------------------------------------------------
+
+  public SimpleHttpResponse postJsonRequest(URI uri, @Nullable String jsonRequestBody)
+      throws IOException {
+    return postJsonRequest(uri, jsonRequestBody, null);
+  }
+
+  public SimpleHttpResponse postJsonRequest(URI uri, @Nullable String requestBody,
+      @Nullable Map<String, String> headers)
+      throws IOException {
+    if (MapUtils.isEmpty(headers)) {
+      headers = new HashMap<>();
+    }
+    headers.put(HttpHeaders.CONTENT_TYPE, JSON_CONTENT_TYPE);
+    HttpEntity entity = requestBody == null ? null : new StringEntity(requestBody, ContentType.APPLICATION_JSON);
+    return sendPostRequest(uri, entity, headers, null);
+  }
+
+  public SimpleHttpResponse putJsonRequest(URI uri, @Nullable String jsonRequestBody)
+      throws IOException {
+    return putJsonRequest(uri, jsonRequestBody, null);
+  }
+
+  public SimpleHttpResponse putJsonRequest(URI uri, @Nullable String requestBody,
+      @Nullable Map<String, String> headers)
+      throws IOException {
+    if (MapUtils.isEmpty(headers)) {
+      headers = new HashMap<>();
+    }
+    headers.put(HttpHeaders.CONTENT_TYPE, JSON_CONTENT_TYPE);
+    HttpEntity entity = requestBody == null ? null : new StringEntity(requestBody, ContentType.APPLICATION_JSON);
+    return sendPutRequest(uri, entity, headers, null);
+  }
+
+  // --------------------------------------------------------------------------
+  // Lower-level request/execute APIs.
+  // --------------------------------------------------------------------------
+
+  public SimpleHttpResponse sendRequest(HttpUriRequest request)
+      throws IOException {
+    try (CloseableHttpResponse response = _httpClient.execute(request)) {
+      String controllerHost = null;
+      String controllerVersion = null;
+      if (response.containsHeader(CommonConstants.Controller.HOST_HTTP_HEADER)) {
+        controllerHost = response.getFirstHeader(CommonConstants.Controller.HOST_HTTP_HEADER).getValue();
+        controllerVersion = response.getFirstHeader(CommonConstants.Controller.VERSION_HTTP_HEADER).getValue();
+      }
+      if (controllerHost != null) {
+        LOGGER.info(String
+            .format("Sending request: %s to controller: %s, version: %s", request.getURI(), controllerHost,
```

Comment:
shall this be extended to talk to server/broker?

Reply:
this is a logging only usage. only Controller response attaches special HTTP response header. but yes we can add others later if necessary